### PR TITLE
bpf: implement unconnected udp based host lb

### PR DESCRIPTION
**Pull request by borkmann**
Rebased & supersedes #8100. /cc @brb

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8375)

### CODE REVIEW

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
Hit #8346

**Comment by borkmann** [comment]
test-me-please

File: bpf/bpf_sock.c [hunk 1]
```diff
@@ -89,15 +138,106 @@ int sock4_xlate(struct bpf_sock_addr *ctx)
 		if (!backend)
 			return CONNECT_PROCEED;
 
+		if (ctx->protocol != IPPROTO_TCP &&
+		    sock4_update_revnat(ctx, backend, &key,
```
Comment (brb) [inline bpf/bpf_sock.c:142]: Maybe it makes sense to log if `map_update_elem()` has failed in `sock4_update_revnat()`. Can be done in a separate PR.

**Review by brb — APPROVED** [review]
Super cool! One nit/question.

File: bpf/bpf_sock.c (hunk 1 above)
Comment (borkmann) [reply bpf/bpf_sock.c:142]: There is currently no perf event output support in the kernel for this type, but we can add it to the kernel. I'll add a separate metric map entry, along with the net-next box update in a next pr.

**Review by joestringer — APPROVED** [review]

**Pull request merged**